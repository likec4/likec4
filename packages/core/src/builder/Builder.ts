import defu from 'defu'
import { fromEntries, hasAtLeast, isFunction, isNonNullish, isNullish, map, mapToObj, mapValues, pickBy } from 'remeda'
import type { Writable } from 'type-fest'
import { invariant } from '../errors'
import { LikeC4Model } from '../model/LikeC4Model'
import {
  type Color,
  type DeploymentRelation,
  type DeploymentView,
  type Element,
  type ElementShape,
  type ElementView,
  type Fqn,
  type IconUrl,
  type LikeC4View,
  type Link,
  type ModelGlobals,
  type ModelRelation,
  type NonEmptyArray,
  type RelationId,
  type Tag,
  DefaultElementShape,
  DefaultThemeColor,
  DeploymentElement,
  isScopedElementView,
} from '../types'
import { isSameHierarchy, nameFromFqn, parentFqn } from '../utils/fqn'
import type { AnyTypes, BuilderSpecification, Types } from './_types'
import type { AddDeploymentNode } from './Builder.deployment'
import type {
  AddDeploymentNodeHelpers,
  DeloymentModelBuildFunction,
  DeloymentModelHelpers,
  DeploymentModelBuilder,
} from './Builder.deploymentModel'
import type { AddElement } from './Builder.element'
import type { AddElementHelpers, ModelBuilder, ModelBuilderFunction, ModelHelpers } from './Builder.model'
import { $autoLayout, $exclude, $include, $rules, $style } from './Builder.view-common'
import type { DeploymentRulesBuilderOp } from './Builder.view-deployment'
import type { ElementViewRulesBuilder } from './Builder.view-element'
import { type ViewsBuilder, type ViewsBuilderFunction, type ViewsHelpers, mkViewBuilder } from './Builder.views'
import type { BuilderMethods } from './Builder.with'

export interface Builder<T extends AnyTypes> extends BuilderMethods<T> {
  /**
   * Only available in compile time
   */
  readonly Types: T

  clone(): Builder<T>

  /**
   * Builders for each element kind
   */
  helpers(): {
    model: ModelHelpers<T>
    views: ViewsHelpers
    deployment: DeloymentModelHelpers<T>
  }

  /**
   * Adds model elements
   *
   * @example
   *  builder.model(({ el }, _) =>
   *    _(
   *      el('a'),
   *      el('a.b').with(
   *        el('c')
   *      )
   *    )
   *  )
   *
   *  builder.model((_, m) =>
   *    m(
   *      _.el('a'),
   *      _.el('a.b').with(
   *        _.el('c')
   *      )
   *    )
   *  )
   */
  model<Out extends AnyTypes>(
    callback: ModelBuilderFunction<T, Out>,
  ): Builder<Out>

  /**
   * Adds deployment model
   *
   * @example
   *  builder.deployment(({ node, instanceOf }, _) =>
   *    _(
   *      node('node1'),
   *      node('node1.child1').with(
   *        instanceOf('model.element')
   *      )
   *    )
   *  )
   *
   *  builder.deployment((_,d) =>
   *    d(
   *      _.node('node1'),
   *       _.node('node1.child1').with(
   *        _.instanceOf('model.element')
   *      )
   *    )
   *  )
   */
  deployment<Out extends AnyTypes>(
    callback: DeloymentModelBuildFunction<T, Out>,
  ): Builder<Out>

  /**
   * Adds views
   *
   * @example
   *  builder.views(({ view, $include }, _) =>
   *    _(
   *      view('index').with(
   *        $include('a -> b')
   *      )
   *    )
   *  )
   */
  views<Out extends AnyTypes>(
    callback: ViewsBuilderFunction<T, Out>,
  ): Builder<Out>

  /**
   * Returns model as result of parsing only
   * Views are not computed or layouted
   * {@link toLikeC4Model} should be used to get model with computed views
   */
  build(): Types.ToParsedLikeC4Model<T>

  /**
   * Returns LikeC4Model with computed views
   */
  toLikeC4Model(): Types.ToLikeC4Model<T>
}

interface Internals<T extends AnyTypes> extends ViewsBuilder<T>, ModelBuilder<T>, DeploymentModelBuilder<T> {
}

type Op<T> = (b: T) => T

function builder<Spec extends BuilderSpecification, T extends AnyTypes>(
  spec: Spec,
  _elements = new Map<string, Element>(),
  _relations = [] as ModelRelation[],
  _views = new Map<string, LikeC4View>(),
  _globals = {
    predicates: {},
    dynamicPredicates: {},
    styles: {},
  } as ModelGlobals,
  _deployments = new Map<string, DeploymentElement>(),
  _deploymentRelations = [] as DeploymentRelation[],
): Builder<T> {
  const toLikeC4Specification = (): Types.ToParsedLikeC4Model<T>['specification'] => ({
    elements: {
      ...structuredClone(spec.elements) as any,
    },
    deployments: {
      ...structuredClone(spec.deployments) as any,
    },
    relationships: {
      ...structuredClone(spec.relationships) as any,
    },
    tags: (spec.tags ?? []) as Tag<T['Tag']>[],
  })

  const mapLinks = (links?: Array<string | { title?: string; url: string }>): NonEmptyArray<Link> | null => {
    if (!links || !hasAtLeast(links, 1)) {
      return null
    }
    return map(links, l => (typeof l === 'string' ? { url: l } : l))
  }

  const createGenericView = <B extends Op<any> | undefined>(
    id: string,
    _props: T['NewViewProps'] | string | B | null,
    builder: B,
  ): [Omit<LikeC4View, 'rules'>, B] => {
    if (isFunction(_props)) {
      builder = _props as B
      _props = {}
    }
    _props ??= {}
    const {
      links: _links = [],
      title = null,
      description = null,
      tags = null,
      ...props
    } = typeof _props === 'string' ? { title: _props } : { ..._props }

    const links = mapLinks(_links)
    return [{
      id: id as any,
      title,
      description,
      tags,
      links,
      customColorDefinitions: {},
      ...props,
    }, builder]
  }

  const self: Builder<T> & Internals<T> = {
    get Types(): T {
      throw new Error('Types are not available in runtime')
    },
    clone: () => {
      return builder(
        structuredClone(spec),
        structuredClone(_elements),
        structuredClone(_relations),
        structuredClone(_views),
        structuredClone(_globals),
        structuredClone(_deployments),
        structuredClone(_deploymentRelations),
      )
    },

    __addElement: (element) => {
      const parent = parentFqn(element.id)
      if (parent) {
        invariant(
          _elements.get(parent),
          `Parent element with id "${parent}" not found for element with id "${element.id}"`,
        )
      }
      if (_elements.has(element.id)) {
        throw new Error(`Element with id "${element.id}" already exists`)
      }
      _elements.set(element.id, element)
      return self
    },
    __addRelation(relation) {
      const sourceEl = _elements.get(relation.source)
      invariant(sourceEl, `Element with id "${relation.source}" not found`)
      const targetEl = _elements.get(relation.target)
      invariant(targetEl, `Element with id "${relation.target}" not found`)
      invariant(
        !isSameHierarchy(sourceEl, targetEl),
        'Cannot create relationship between elements in the same hierarchy',
      )
      _relations.push({
        id: `rel${_relations.length + 1}` as RelationId,
        ...relation,
      })
      return self
    },
    /**
     * Fully qualified name for nested elements
     */
    __fqn(id) {
      invariant(id.trim() !== '', 'Id must be non-empty')
      return id as Fqn
    },
    __addSourcelessRelation() {
      throw new Error('Can be called only in nested model')
    },
    __addView: (view) => {
      if (_views.has(view.id)) {
        throw new Error(`View with id "${view.id}" already exists`)
      }
      if (isScopedElementView(view)) {
        invariant(
          _elements.get(view.viewOf),
          `Invalid scoped view ${view.id}, wlement with id "${view.viewOf}" not found`,
        )
      }
      _views.set(view.id, view)
      return self
    },

    __addDeployment: (node: DeploymentElement) => {
      if (_deployments.has(node.id)) {
        throw new Error(`Deployment with id "${node.id}" already exists`)
      }
      const parent = parentFqn(node.id)
      if (parent) {
        invariant(
          _deployments.get(parent),
          `Parent element with id "${parent}" not found for node with id "${node.id}"`,
        )
      }
      if (DeploymentElement.isInstance(node)) {
        invariant(parent, `Instance ${node.id} of ${node.element} must be deployed under a parent node`)
        invariant(
          _elements.get(node.element),
          `Instance "${node.id}" references non-existing element "${node.element}"`,
        )
      }
      _deployments.set(node.id, node)
      return self
    },
    __addDeploymentRelation: (relation) => {
      invariant(
        !isSameHierarchy(relation.source.id, relation.target.id),
        'Cannot create relationship between elements in the same hierarchy',
      )

      invariant(
        _deployments.has(relation.source.id),
        `Relation "${relation.source.id} -> ${relation.target.id}" references non-existing source`,
      )
      invariant(
        _deployments.has(relation.target.id),
        `Relation "${relation.source.id} -> ${relation.target.id}" references non-existing target`,
      )

      _deploymentRelations.push({
        id: `deploy_rel${_deploymentRelations.length + 1}` as RelationId,
        ...relation,
      })
      return self
    },
    build: () => ({
      specification: toLikeC4Specification(),
      elements: fromEntries(
        structuredClone(
          Array.from(_elements.entries()),
        ),
      ),
      relations: mapToObj(_relations, r => [r.id, structuredClone(r)]),
      globals: structuredClone(_globals),
      deployments: {
        elements: fromEntries(
          structuredClone(
            Array.from(_deployments.entries()),
          ),
        ),
        relations: mapToObj(_deploymentRelations, r => [r.id, structuredClone(r)]),
      },
      views: fromEntries(
        structuredClone(
          Array.from(_views.entries()),
        ),
      ),
    } as any),
    toLikeC4Model: () => {
      const parsed = self.build()
      return LikeC4Model.compute(parsed)
    },
    helpers: () => ({
      model: {
        model: (...ops: ((b: ModelBuilder<T>) => ModelBuilder<T>)[]) => {
          return (b: Builder<T>) => {
            return ops.reduce((b, op) => op(b), b as any as ModelBuilder<T>) as any
          }
        },
        rel: (source: string, target: string, _props?: T['NewRelationshipProps'] | string) => {
          return <T extends AnyTypes>(b: ModelBuilder<T>) => {
            const {
              title = '',
              links: _links = [],
              ...props
            } = defu(
              typeof _props === 'string' ? { title: _props } : { ..._props },
              { title: null, links: null },
            )
            const links = mapLinks(_links)
            b.__addRelation({
              source: source as any,
              target: target as any,
              title,
              ...(links && { links }),
              ...props,
            })
            return b
          }
        },
        relTo: (target, _props?) => {
          return <T extends AnyTypes>(b: ModelBuilder<T>) => {
            const {
              title = '',
              links: _links = [],
              ...props
            } = defu(
              typeof _props === 'string' ? { title: _props } : { ..._props },
              { title: null, links: null },
            )
            const links = mapLinks(_links)
            b.__addSourcelessRelation({
              target,
              title,
              ...(links && { links }),
              ...props,
            })
            return b
          }
        },
        ...mapValues(
          spec.elements,
          ({ style: specStyle, ...spec }, kind) =>
          <Id extends string>(
            id: Id,
            _props?: T['NewElementProps'] | string,
          ): AddElement<Id> => {
            const add = (<T extends AnyTypes>(b: ModelBuilder<T>) => {
              const {
                links: _links,
                icon: _icon,
                style,
                title,
                ...props
              } = typeof _props === 'string' ? { title: _props } : { ..._props }

              const links = mapLinks(_links)

              const icon = _icon ?? specStyle?.icon

              const _id = b.__fqn(id)

              b.__addElement({
                id: _id,
                kind: kind as any,
                title: title ?? nameFromFqn(_id),
                description: null,
                technology: null,
                tags: null,
                color: specStyle?.color ?? DefaultThemeColor as Color,
                shape: specStyle?.shape ?? DefaultElementShape as ElementShape,
                style: pickBy({
                  border: specStyle?.border,
                  opacity: specStyle?.opacity,
                  size: specStyle?.size,
                  padding: specStyle?.padding,
                  textSize: specStyle?.textSize,
                  ...style,
                }, isNonNullish),
                links,
                ...icon && { icon: icon as IconUrl },
                ...spec,
                ...props,
              })
              return b
            }) as AddElement<Id>

            add.with = (...ops: Array<(input: ModelBuilder<any>) => ModelBuilder<any>>) => (b: ModelBuilder<any>) => {
              add(b)
              const { __fqn, __addSourcelessRelation } = b
              try {
                b.__fqn = (child) => `${__fqn(id)}.${child}` as Fqn
                b.__addSourcelessRelation = (relation) => {
                  return b.__addRelation({
                    ...relation,
                    source: __fqn(id),
                  })
                }
                ops.reduce((b, op) => op(b), b as any as ModelBuilder<T>) as any
              } finally {
                b.__fqn = __fqn
                b.__addSourcelessRelation = __addSourcelessRelation
              }
              return b
            }

            return add
          },
        ) as AddElementHelpers<T>,
      },
      views: {
        views: (...ops: ((b: ViewsBuilder<T>) => ViewsBuilder<T>)[]) => {
          return (b: Builder<T>) => {
            return ops.reduce((b, op) => op(b), b as any as ViewsBuilder<T>) as any
          }
        },
        view: (
          id: string,
          _props?: T['NewViewProps'] | string | ElementViewRulesBuilder<any>,
          _builder?: ElementViewRulesBuilder<any>,
        ) => {
          const [generic, builder] = createGenericView(id, _props, _builder)
          const view: Writable<ElementView> = {
            ...generic,
            __: 'element',
            rules: [],
          }

          const add = (b: ViewsBuilder<any>): ViewsBuilder<any> => {
            b.__addView(view)
            if (builder) {
              builder(mkViewBuilder(view))
            }
            return b
          }
          add.with = (...ops: ElementViewRulesBuilder<any>[]) => (b: ViewsBuilder<any>) => {
            add(b)
            const elementViewBuilder = mkViewBuilder(view)
            for (const op of ops) {
              op(elementViewBuilder)
            }
            return b
          }

          return add
        },
        viewOf: (
          id,
          viewOf: string,
          _props?: T['NewViewProps'] | string | ElementViewRulesBuilder<any>,
          _builder?: ElementViewRulesBuilder<any>,
        ) => {
          const [generic, builder] = createGenericView(id, _props, _builder)
          const view: Writable<ElementView> = {
            ...generic,
            viewOf: viewOf as Fqn,
            __: 'element',
            rules: [],
          }

          const add = (b: ViewsBuilder<any>): ViewsBuilder<any> => {
            b.__addView(view)
            if (builder) {
              builder(mkViewBuilder(view))
            }
            return b
          }

          add.with = (...ops: ElementViewRulesBuilder<any>[]) => (b: ViewsBuilder<any>) => {
            add(b)
            const elementViewBuilder = mkViewBuilder(view)
            for (const op of ops) {
              op(elementViewBuilder)
            }
            return b
          }

          return add
        },

        deploymentView: (
          id: string,
          _props?: T['NewViewProps'] | string | DeploymentRulesBuilderOp<any>,
          _builder?: DeploymentRulesBuilderOp<any>,
        ) => {
          const [generic, builder] = createGenericView(id, _props, _builder)
          const view: Writable<DeploymentView> = {
            ...generic,
            __: 'deployment',
            rules: [],
          }

          const add = (b: ViewsBuilder<any>): ViewsBuilder<any> => {
            b.__addView(view)
            if (builder) {
              builder(mkViewBuilder(view))
            }
            return b
          }

          add.with = (...ops: DeploymentRulesBuilderOp<any>[]) => (b: ViewsBuilder<any>) => {
            add(b)
            const elementViewBuilder = mkViewBuilder(view)
            for (const op of ops) {
              op(elementViewBuilder)
            }
            return b
          }

          return add
        },
        $autoLayout,
        $exclude,
        $include,
        $rules,
        $style,
      },
      deployment: {
        deployment: (...ops: ((b: DeploymentModelBuilder<T>) => DeploymentModelBuilder<T>)[]) => {
          return (b: Builder<T>) => {
            return ops.reduce((b, op) => op(b), b as any as DeploymentModelBuilder<T>) as any
          }
        },
        instanceOf: (
          id: string,
          target?: string | T['NewDeploymentNodeProps'],
          _props?: string | T['NewDeploymentNodeProps'],
        ) => {
          return <T extends AnyTypes>(
            b: DeploymentModelBuilder<T>,
          ): DeploymentModelBuilder<Types.AddDeploymentFqn<T, any>> => {
            if (isNullish(target)) {
              target = id
              id = nameFromFqn(id)
            } else if (typeof target === 'string') {
              _props ??= {}
            } else {
              _props = target
              target = id
              id = nameFromFqn(id)
            }
            const {
              links,
              title,
              ...props
            } = typeof _props === 'string' ? { title: _props } : { ..._props }
            const _id = b.__fqn(id)
            invariant(_elements.has(target), `Target element with id "${target}" not found`)
            b.__addDeployment({
              id: _id,
              element: target as any,
              ...title && { title },
              ...links && { links: mapLinks(links) },
              ...props,
            })
            return b as any
          }
        },
        rel: (source: string, target: string, _props?: T['NewRelationshipProps'] | string) => {
          return <T extends AnyTypes>(b: DeploymentModelBuilder<T>) => {
            const {
              title,
              links,
              ...props
            } = typeof _props === 'string' ? { title: _props } : { ..._props }

            b.__addDeploymentRelation({
              source: {
                id: source as any,
              },
              target: {
                id: target as any,
              },
              ...title && { title },
              ...links && { links: mapLinks(links) },
              ...props,
            })
            return b
          }
        },
        ...mapValues(
          spec.deployments ?? {},
          ({ style: specStyle, ...spec }, kind) =>
          <Id extends string>(
            id: Id,
            _props?: T['NewDeploymentNodeProps'] | string,
          ): AddDeploymentNode<Id> => {
            const add = (<T extends AnyTypes>(b: DeploymentModelBuilder<T>) => {
              const {
                links,
                icon: _icon,
                style,
                title,
                ...props
              } = typeof _props === 'string' ? { title: _props } : { ..._props }

              const icon = _icon ?? specStyle?.icon

              const _id = b.__fqn(id)

              b.__addDeployment({
                id: _id,
                kind: kind as any,
                title: title ?? nameFromFqn(_id),
                description: null,
                technology: null,
                tags: null,
                color: specStyle?.color ?? DefaultThemeColor as Color,
                shape: specStyle?.shape ?? DefaultElementShape as ElementShape,
                style: pickBy({
                  border: specStyle?.border,
                  opacity: specStyle?.opacity,
                  size: specStyle?.size,
                  padding: specStyle?.padding,
                  textSize: specStyle?.textSize,
                  ...style,
                }, isNonNullish),
                ...links && { links: mapLinks(links) },
                ...icon && { icon: icon as IconUrl },
                ...spec,
                ...props,
              })
              return b
            }) as AddDeploymentNode<Id>

            add.with =
              (...ops: Array<(input: DeploymentModelBuilder<any>) => DeploymentModelBuilder<any>>) =>
              (b: DeploymentModelBuilder<any>) => {
                add(b)
                const { __fqn } = b
                try {
                  b.__fqn = (child) => `${__fqn(id)}.${child}` as Fqn
                  ops.reduce((b, op) => op(b), b as any as DeploymentModelBuilder<T>) as any
                } finally {
                  b.__fqn = __fqn
                }
                return b
              }
            return add
          },
        ) as AddDeploymentNodeHelpers<T>,
      },
    } as {
      model: ModelHelpers<T>
      views: ViewsHelpers
      deployment: DeloymentModelHelpers<T>
    }),
    with: (...ops: ((b: Builder<T>) => Builder<T>)[]) => {
      return ops.reduce((b, op) => op(b), self as Builder<T>).clone()
    },
    model: <Out extends AnyTypes>(cb: ModelBuilderFunction<T, Out>) => {
      const b = self.clone()
      const helpers = b.helpers().model
      const _ = helpers.model as any
      return cb({ ...helpers, _ }, _)(b as Internals<T>) as any
    },
    deployment: <Out extends AnyTypes>(cb: DeloymentModelBuildFunction<T, Out>) => {
      const b = self.clone()
      const helpers = b.helpers().deployment
      const _ = helpers.deployment as any
      return cb({ ...helpers, _ }, _)(b as Internals<T>) as any
    },
    views: <Out extends AnyTypes>(cb: ViewsBuilderFunction<T, Out>) => {
      const b = self.clone()
      const helpers = b.helpers().views
      return cb({
        ...helpers,
        _: helpers.views as any,
      } as any, helpers.views as any)(b as Internals<T>) as any
    },
  }

  return self
}

export namespace Builder {
  export type Any = Builder<AnyTypes>

  export function forSpecification<const Spec extends BuilderSpecification>(
    spec: Spec,
  ): {
    builder: Builder<Types.FromSpecification<Spec>>
    model: ModelHelpers<Types.FromSpecification<Spec>>
    deployment: DeloymentModelHelpers<Types.FromSpecification<Spec>>
    views: ViewsHelpers
  } {
    const b = builder<Spec, Types.FromSpecification<Spec>>(spec)
    return {
      ...b.helpers(),
      builder: b,
    }
  }

  export function specification<const Spec extends BuilderSpecification>(
    spec: Spec,
  ): Builder<Types.FromSpecification<Spec>> {
    return builder<Spec, Types.FromSpecification<Spec>>(spec)
  }
}
