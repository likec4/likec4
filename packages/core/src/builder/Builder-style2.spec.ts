import { describe, expect, it } from 'vitest'
import { Builder } from './Builder'

describe('Builder (style 2)', () => {
  const spec = Builder
    .specification({
      elements: {
        system: {
          style: {
            size: 'lg',
            textSize: 'sm',
          },
        },
        component: {},
        actor: {},
      },
      deployments: {
        env: {
          style: {
            size: 'lg',
          },
        },
        node: {},
      },
    })

  it('should build ', () => {
    const b = spec.clone()
      .model(({ system, actor, component }, _) =>
        _(
          actor('customer'),
          system('cloud').with(
            component('ui'),
          ),
        )
      )
      .deployment(({ env, node, instanceOf }, _) =>
        _(
          env('prod').with(
            node('eu').with(
              instanceOf('cloud.ui'),
            ),
          ),
          env('dev'),
          node('dev.local'),
        )
      )
      .views(({ view, viewOf, deploymentView, $include }, _) =>
        _(
          view('index', 'Index').with(
            $include('cloud.*'),
          ),
          viewOf('cloud', 'cloud.ui').with(
            $include('* -> cloud.**'),
          ),
          deploymentView('deployment', 'Deployment').with(
            $include('prod.**'),
          ),
        )
      )

    expect(b.build()).toMatchSnapshot()
  })

  it('should fail if invalid ID provided ', () => {
    expect(() => {
      spec.model(({ actor }, _) =>
        _(
          actor('cust.omer'),
        )
      )
    }).toThrowError('Parent element with id "cust" not found for element with id "cust.omer"')
  })

  it('should fail on invalid instance ', () => {
    const b = spec.clone()
      .model(_ =>
        _.model(
          _.component('cloud'),
          _.component('cloud.ui'),
        )
      )

    expect(() => {
      const raw = b.deployment(_ =>
        _.deployment(
          _.instanceOf('cloud.ui'),
        )
      ).build()
    }).toThrowError('Instance ui of cloud.ui must be deployed under a parent node')

    // Nested instanceOf is correct

    const raw = b.deployment(_ =>
      _.deployment(
        _.node('node').with(
          _.instanceOf('cloud.ui'),
        ),
      )
    ).build()
    expect(raw.deployments.elements).toEqual({
      node: expect.objectContaining({ id: 'node' }),
      // Take name from element
      'node.ui': expect.objectContaining({ id: 'node.ui', element: 'cloud.ui' }),
    })
  })
})
