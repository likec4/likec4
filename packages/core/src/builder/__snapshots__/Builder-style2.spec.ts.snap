// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Builder (style 2) > should build  1`] = `
{
  "deployments": {
    "elements": {
      "dev": {
        "color": "primary",
        "description": null,
        "id": "dev",
        "kind": "env",
        "shape": "rectangle",
        "style": {
          "size": "lg",
        },
        "tags": null,
        "technology": null,
        "title": "dev",
      },
      "dev.local": {
        "color": "primary",
        "description": null,
        "id": "dev.local",
        "kind": "node",
        "shape": "rectangle",
        "style": {},
        "tags": null,
        "technology": null,
        "title": "local",
      },
      "prod": {
        "color": "primary",
        "description": null,
        "id": "prod",
        "kind": "env",
        "shape": "rectangle",
        "style": {
          "size": "lg",
        },
        "tags": null,
        "technology": null,
        "title": "prod",
      },
      "prod.eu": {
        "color": "primary",
        "description": null,
        "id": "prod.eu",
        "kind": "node",
        "shape": "rectangle",
        "style": {},
        "tags": null,
        "technology": null,
        "title": "eu",
      },
      "prod.eu.ui": {
        "element": "cloud.ui",
        "id": "prod.eu.ui",
      },
    },
    "relations": {},
  },
  "elements": {
    "cloud": {
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "shape": "rectangle",
      "style": {
        "size": "lg",
        "textSize": "sm",
      },
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    "cloud.ui": {
      "color": "primary",
      "description": null,
      "id": "cloud.ui",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "ui",
    },
    "customer": {
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {
    "cloud": {
      "__": "element",
      "customColorDefinitions": {},
      "description": null,
      "id": "cloud",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "source": {
                "wildcard": true,
              },
              "target": {
                "element": "cloud",
                "isDescendants": true,
              },
            },
          ],
        },
      ],
      "tags": null,
      "title": null,
      "viewOf": "cloud.ui",
    },
    "deployment": {
      "__": "deployment",
      "customColorDefinitions": {},
      "description": null,
      "id": "deployment",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "ref": {
                "deployment": "prod",
              },
              "selector": "descendants",
            },
          ],
        },
      ],
      "tags": null,
      "title": "Deployment",
    },
    "index": {
      "__": "element",
      "customColorDefinitions": {},
      "description": null,
      "id": "index",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "element": "cloud",
              "isChildren": true,
            },
          ],
        },
      ],
      "tags": null,
      "title": "Index",
    },
  },
}
`;
